lifecycle-coroutines/src/commonMain/kotlin/com/arkivanov/essenty/lifecycle/coroutines/LifecycleCoroutinesExt.kt

```diff
@@ -0,0 +1,147 @@
+) {
+    require(state != Lifecycle.State.INITIALIZED) {
+        "repeatOnEssentyLifecycle cannot start work with the INITIALIZED lifecycle state."
+    }
+
+    if (this.state == Lifecycle.State.DESTROYED) return
+
+    withCoroutineScope {
+        if (this@repeatOnEssentyLifecycle.state == Lifecycle.State.DESTROYED) return@withCoroutineScope
+
+        var callback: Lifecycle.Callbacks? = null
+        var job: Job? = null
+        val mutex = Mutex()
+
+        try {
+            suspendCancellableCoroutine { cont ->
+                callback = createLifecycleAwareCallback(
+                    startState = state,
+                    onStateAppear = {
+                        job = startJob(mutex, block)
+                    },
+                    onStateDisappear = {
+                        job?.cancel()
+                        job = null
+                    },
+                    onDestroy = { cont.resume(Unit) },
+                )
+
+                this@repeatOnEssentyLifecycle.subscribe(callback as Lifecycle.Callbacks)
+            }
+        } finally {
+            job?.cancel()
+            job = null
+            callback?.let {
+                this@repeatOnEssentyLifecycle.unsubscribe(it)
+            }
+        }
+    }
+}
+
+/**
+ * Start [Flow] collecting when [Lifecycle.State] is at least as [minActiveState].
+ * It stopped collecting if "opposite" [Lifecycle.State] will appear.
+ */
+fun <T> Flow<T>.flowWithEssentyLifecycle(
+    lifecycle: Lifecycle,
+    minActiveState: Lifecycle.State = Lifecycle.State.STARTED
+): Flow<T> = callbackFlow {
+    lifecycle.repeatOnEssentyLifecycle(minActiveState) {
+        this@flowWithEssentyLifecycle.collect {
+            send(it)
+        }
+    }
+    close()
+}
+
+/**
+ * Function for executing [block] on the [CoroutineScope] with passed specific [dispatcher].
+ */
+private suspend fun withCoroutineScope(
+    dispatcher: CoroutineDispatcher = Dispatchers.Main.immediate,
+    block: suspend CoroutineScope.() -> Unit
+): Unit = coroutineScope {
+    withContext(dispatcher) {
+        block()
+    }
+}
+
+/**
+ * Creates lifecycle aware [Lifecycle.Callbacks] interface instance.
+ *
+ * @param startState [Lifecycle.State] that [onStateAppear] block must be called from
+ * @param onStateAppear block of code that will be executed when the [Lifecycle.State] was equal [startState]
+ * @param onStateDisappear block of code that will be executed when the [Lifecycle.State] was equal to opposite [startState]
+ * @param onDestroy block of code that will be executed when the [Lifecycle.State] was equal [Lifecycle.State.DESTROYED]
+ *
+ * @return [Lifecycle.Callbacks]
+ */
+private fun createLifecycleAwareCallback(
+    startState: Lifecycle.State,
+    onStateAppear: () -> Unit,
+    onStateDisappear: () -> Unit,
+    onDestroy: () -> Unit,
+): Lifecycle.Callbacks = object : Lifecycle.Callbacks {
+
+    override fun onCreate(): Unit = launchWithState(Lifecycle.State.CREATED)
```

Comment:
Could you please use method bodies for methods returning `Unit`? Similarly to how you did with `onDestroy`.

Reply:
Ok, replaced to method bodies.